File: apps/fsm-core-py-fastapi/app/workers/fsmpromiseworker.py
import asyncio
import logging
from typing import Optional
import asyncpg

from app.db.queue import read_message
from app.workers.fsmpromiseworker_helper import process_fsm_promise_queue_message

logger = logging.getLogger(__name__)
VISIBILITY_TIMEOUT = 30


async def start_fsm_promise_worker(
    pool: asyncpg.Pool,
    queue_name: str,
    fsm_promise_name: str,
    fsm_promise_version: Optional[str] = None,
) -> None:
    logger.info(
        f"FSM Promise worker started | queue={queue_name} promise={fsm_promise_name} version={fsm_promise_version}"
    )

    while True:
        messages = await read_message(pool, queue_name, VISIBILITY_TIMEOUT)
        if not messages:
            await asyncio.sleep(1)
            continue

        for msg in messages:
            if not (msg.get("message") and msg.get("msg_id")):
                continue
            try:
                await process_fsm_promise_queue_message(
                    pool,
                    queue_name,
                    msg,
                    fsm_promise_name,
                    str(fsm_promise_version) if fsm_promise_version else None,
                )
            except Exception:
                logger.exception(
                    f"Error processing FSM promise message in queue '{queue_name}'"
                )

File: apps/fsm-core-ts-hono-deno/worker/fsmpromiseworker.ts
import type { Database } from "../../fsm-core-db-ts/src/database.types.ts";

import type { DBDeps } from "../../fsm-core-db-ts/src/custom-type.ts";

import {
  readMessage,
} from "../../fsm-core-db-ts/src/queue.ts";

import { processFSMPromiseQueueMessage } from "./fsmpromiseworker-helper.ts";

export async function startFSMPromiseWorker(
  deps: DBDeps,
  queueName: string,
  fsm_promise_name: string,
  fsm_promise_version: number,
) {
  const visibilityTimeout = 30;
  console.log(`👷 Started FSM Promise worker for queue: ${queueName}`);

  while (true) {
    const messages = await readMessage(deps, queueName, visibilityTimeout);
    if (messages.length === 0) {
      await new Promise((resolve) => setTimeout(resolve, 1000));
      continue;
    }

    for (const msg of messages) {
      if (msg.message && msg.msg_id) {
        try {
          console.log("✅ Processing FSM promise message:", msg.message);
          // Here you would process the message
          await processFSMPromiseQueueMessage(
            deps,
            queueName,
            msg,
            fsm_promise_name,
            fsm_promise_version.toString(),
          );

          // await archiveMessage(deps, queueName, msg.msg_id || 1);
        } catch (err) {
          console.error("❌ Error processing FSM promise message:", err);
        }
      }
    }
  }
}
